multimodal/src/autogluon/multimodal/learners/default_learner.py

```diff
@@ -0,0 +1,1724 @@
+        save_path: str,
+        ckpt_path: str,  # these two are synced
+        resume: bool,  # these two are synced
+        max_time: Optional[timedelta] = None,
+        enable_progress_bar: Optional[bool] = False,  # can be inferred?
+        presets: Optional[str] = None,
+        config: Optional[dict] = None,
+        hyperparameters: Optional[Union[str, Dict, List[str]]] = None,
+        advanced_hyperparameters: Optional[Dict] = None,
+        teacher_predictor: Union[str, DefaultLearner] = None,
+        hpo_mode: bool = False,
+        standalone: bool = True,
+        clean_ckpts: bool = True,
+        **hpo_kwargs,
+    ):
+        # TODO(?) We should have a separate "_pre_training_event()" for logging messages.
+        logger.info(get_fit_start_message(save_path, validation_metric_name))
+
+        ### SETUP Environments ###
+        self._fit_called = True
+        # 1-3, 22. get config.
+        config, df_preprocessor, grad_steps, strategy, use_ray_lightning = self._setup_environment(
+            train_df, config, presets, hyperparameters, teacher_predictor, hpo_mode, **hpo_kwargs
+        )
+
+        # 4. if NER, update output shape. NOTE: This can be refactored into the NER Learner
+        # Update output_shape with label_generator.
+        model = self._create_model(config, df_preprocessor)
+
+        # 6. get trainable layer params for efficient finetuning only?
+        trainable_param_names = self._get_trainable_params(config, model)
+
+        # 7. get data_processors.
+        data_processors = self._get_data_processors(config, advanced_hyperparameters, model)
+
+        # 8. infer positive labels (NOTE: This is deleted in the latest main branch)
+        # pos_label = try_to_infer_pos_label(
+        #     data_config=config.data,
+        #     label_encoder=df_preprocessor.label_generator,
+        #     problem_type=self._problem_type,
+        # )
+
+        # 9. setup validation metric
+        validation_metric, custom_metric_func = self._setup_validation_metric(validation_metric_name)
+
+        # 10. setup mix up if applicable
+        mixup_active, mixup_fn = self._setup_mixup(config)
+
+        # 11. get loss function, NOTE: This can be refactored into Learner class
+        loss_func = get_loss_func(
+            problem_type=self._problem_type,
+            mixup_active=mixup_active,
+            loss_func_name=OmegaConf.select(config, "optimization.loss_function"),
+            config=config.optimization,
+        )
+
+        # 12. get post process function
+        model_postprocess_fn = get_model_postprocess_fn(
+            problem_type=self._problem_type,
+            loss_func=loss_func,
+        )
+
+        # 13. assign properties
+        self._config = config
+        self._df_preprocessor = df_preprocessor
+        self._data_processors = data_processors
+        self._model = model
+        self._model_postprocess_fn = model_postprocess_fn
```

Comment:
what happens if this isn't the first call to `fit`? The state seems to start getting pretty complicated at that point.

Reply:
If there isn't a first call, then these properties will be set to None. For other methods like `.evaluate()`, `.predict()`, learner checks to see if it is ready for inference by using these states.